hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenMintHandler.java

```diff
@@ -54,8 +77,169 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
+                    context.consensusNow(),
+                    accountStore,
+                    tokenStore,
+                    tokenRelStore,
+                    nftStore);
+            final var recordBuilder = context.recordBuilder(TokenMintRecordBuilder.class);
+
+            recordBuilder.serialNumbers(mintedSerials);
+            // TODO: Need to build transfer ownership from list to transfer NFT to treasury
+        }
+    }
+
+    /**
+     * Validates the semantics of the token mint transaction that involve state or config.
+     * @param context - the handle context of the token mint transaction
+     */
+    private void validateSemantics(final HandleContext context) {
+        requireNonNull(context);
+        final var op = context.body().tokenMintOrThrow();
+        validator.validate(op.amount(), op.metadata());
+    }
+
+    /**
+     * Minting nfts creates new instances of the given non-fungible token. Increments the
+     * serial number of the given base unique token, and increments total owned nfts of the
+     * non-fungible token.
+     *
+     * @param token
+     * @param treasuryRel   - the treasury relation of the token
+     * @param metadata      - the metadata of the nft to be minted
+     * @param consensusTime - the consensus time of the transaction
+     * @param accountStore  - the account store
+     * @param tokenStore    - the token store
+     * @param tokenRelStore - the token relation store
+     * @param nftStore      - the nft store
+     */
+    private List<Long> mintNonFungible(
+            final Token token,
+            @NonNull final TokenRelation treasuryRel,
+            @NonNull final List<Bytes> metadata,
+            @NonNull final Instant consensusTime,
+            @NonNull final WritableAccountStore accountStore,
+            @NonNull final WritableTokenStore tokenStore,
+            @NonNull final WritableTokenRelationStore tokenRelStore,
+            @NonNull final WritableNftStore nftStore) {
+        final var metadataCount = metadata.size();
+        validateFalse(metadata.isEmpty(), INVALID_TOKEN_MINT_METADATA);
+
+        // validate token number from treasury relation
+        final var tokenId = asToken(treasuryRel.tokenNumber());
+        validateTrue(treasuryRel.tokenNumber() == token.tokenNumber(), FAIL_INVALID);
+
+        // get the treasury account
+        final var treasuryAccount = accountStore.get(asAccount(treasuryRel.accountNumber()));
+        validateTrue(treasuryAccount != null, INVALID_TREASURY_ACCOUNT_FOR_TOKEN);
+
+        // get the latest serial number minted for the token
+        var currentSerialNumber = token.lastUsedSerialNumber();
+        validateTrue((currentSerialNumber + metadataCount) <= MAX_NUM_ALLOWED, SERIAL_NUMBER_LIMIT_REACHED);
+
+        // Change the supply on token
+        changeSupply(token, treasuryRel, metadataCount, FAIL_INVALID, accountStore, tokenStore, tokenRelStore);
+
+        final var mintedSerials = new ArrayList<Long>();
```

Comment:
This could be instantiated with the size of the metadata list to avoid resizing, but again not a big deal

```java
new ArrayList<Long>(metadata.size());
```

Reply:
Done